### PR TITLE
Implement full login flow

### PR DESCRIPTION
**Pull request by japorized**
This PR implements the full login flow from the web frontend to the API backend.

There's not much of a surprise here.

Essentially, the flow goes like:

1. User visits `web/login`
2. Gets redirected `api/login`
3. Gets redirected to Github for OAuth2.0 authentication
4. Gets redirected back to `api/login/callback`
5. Gets redirected back to `web/login/callback`
6. Gets redirected to either `web/` or the hinted web uri

### CODE REVIEW

File: api/controllers/login.go [hunk 1]
```diff
@@ -73,8 +101,24 @@ func (ctrl *Controller) OAuthCallbackHandler(c *gin.Context) {
 		return
 	}
 
-	c.JSON(http.StatusOK, gin.H{
-		"access_token": userAuthToken.Token,
-		"expires_at":   userAuthToken.ExpiresAt,
-	})
+	c.Header("Set-Cookie", fmt.Sprintf("Access-Token=%s; Expires=%s; HttpOnly; Secure; SameSite=Lax; Path=/", userAuthToken.Token, userAuthToken.ExpiresAt.Format(http.TimeFormat)))
```
Comment (mosiman) [inline api/controllers/login.go:104]: Is "Access Token" standard, or should we make it look salz-specific? e.g. `Salz-Access-Token`?

Comment (mosiman) [reply api/controllers/login.go:104]: its a nitpick though, doesn't really matter to me

**Review by mosiman — APPROVED** [review]
i'd be fine if this merged, but i did have some comments and questions.

I checked it out and it seems to work (mostly).

- I think we need to be more specific about the cors origins. I'm getting the error in the console: `Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at ‘http://localhost:8080/users/me’. (Reason: Credential is not supported if the CORS header ‘Access-Control-Allow-Origin’ is ‘*’).`. 
- That being said, we can open an API ticket for it later since it seems to work anyways: I see the cookie set with the token. 
image

File: web/src/routes/(all_pages)/settings/+page.svelte [hunk 2]
```diff
@@ -25,6 +28,27 @@
...
+				return;
+			}
+			getHealth(settingsClone.apiBaseurl).then((resp) => {
+				isApiOnline = resp.isOk() ? 'online' : 'offline';
+			});
+		} else {
+			isApiOnline = 'unknown';
+		}
+	};
+
```
[the remark this answers is not in the document]
Comment (mosiman) [reply web/src/routes/(all_pages)/settings/+page.svelte:51]: oh but the comment indents are weird

image

File: api/controllers/login.go [hunk 3]
```diff
@@ -15,14 +19,38 @@ const (
...
-	// TODO Check the state to prevent CSRF attacks
+	// TODO Maybe redo encodedState message to do checks to prevent CSRF attacks?
+	encodedState := c.Query("state")
+	var state State
+	if err := json.Unmarshal([]byte(encodedState), &state); err != nil {
+		log.Error("Could not parse state", "error", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+	log.Info("Print state", "state", state)
```
[the remark this answers is not in the document]
Comment (japorized) [reply api/controllers/login.go:52]: Meant to nuke that. This is what happens when you leave the codebase in a state of debugging and come back after a few days xD

File: api/controllers/login.go [hunk 4]
```diff
@@ -73,8 +101,24 @@ func (ctrl *Controller) OAuthCallbackHandler(c *gin.Context) {
...
+	if err != nil {
+		log.Error("Could not create redirect url to web", "error", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+	redirectUrl.URL.RawQuery = url.Values{
+		"redirect_uri": {state.RedirectUri},
+	}.Encode()
+
+	c.Redirect(http.StatusTemporaryRedirect, redirectUrl.URL.String())
```
[the remark this answers is not in the document]
Comment (japorized) [reply api/controllers/login.go:123]: Oh, good point. We can do that. Force the frontend to always pass some redirect url.

File: web/src/lib/req/index.ts [hunk 5]
```diff
@@ -7,7 +6,7 @@ export type Resp<T extends Record<string, unknown>> = {
 };
 
 export const req = async <T extends Record<string, unknown>>(
-	method: HttpMethod,
+	method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'OPTION',
```
[the remark this answers is not in the document]
Comment (japorized) [reply web/src/lib/req/index.ts:9]: This was done a long time ago and `HttpMethod` doesn't have `'OPTION'` I think, which is why I decided to just hand-wrung the methods here.

File: web/src/routes/(all_pages)/login/callback/+page.svelte [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	import { settingsStore } from '$lib/stores/settings';
+	import { userStore, isLoggedIn } from '$lib/stores/user';
+	import { onMount } from 'svelte';
+
+	onMount(async () => {
+		if (isLoggedIn) {
+			goto('/');
+		}
+
+		let redirectUrl = decodeURIComponent($page.url.searchParams.get('redirect_uri') ?? '%2F');
```
[the remark this answers is not in the document]
Comment (japorized) [reply web/src/routes/(all_pages)/login/callback/+page.svelte:14]: The old web dev in me is mad at the fact that people don't remember common urlencoded characters xD

File: web/src/routes/(all_pages)/settings/+page.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (japorized) [reply web/src/routes/(all_pages)/settings/+page.svelte:51]: This gets removed in the next PR cause I decided that it's not a good idea to just try ping the server on a loop.

**Review by mosiman — APPROVED** [review]

**Review by Lugarun — APPROVED** [review]
image

**Pull request merged**